src/theory/symmetry_detect.h

```diff
@@ -0,0 +1,146 @@
+/*********************                                                        */
+/*! \file symmetry_detect.h
+ ** \verbatim
+ ** Top contributors (to current version):
+ **  Paul Meng, Andrew Reynolds
+ ** This file is part of the CVC4 project.
+ ** Copyright (c) 2009-2018 by the authors listed in the file AUTHORS
+ ** in the top-level source directory) and their institutional affiliations.
+ ** All rights reserved.  See the file COPYING in the top-level source
+ ** directory for licensing information.\endverbatim
+ **
+ ** \brief Symmetry detection for theories
+ **/
+
+#include "cvc4_private.h"
+
+#ifndef __CVC4__THEORY__SYMMETRY_DETECT_H
+#define __CVC4__THEORY__SYMMETRY_DETECT_H
+
+#include <map>
+#include <string>
+#include <vector>
+#include "expr/node.h"
+
+namespace CVC4 {
+
+/**
+ * This is the class to detect symmetries from input based on terms equality.
+ * */
+class SymmetryDetect
+{
+ public:
+  /**
+   * Constructor
+   * */
+  SymmetryDetect()
+  {
+    d_trueNode = NodeManager::currentNM()->mkConst<bool>(true);
+    d_falseNode = NodeManager::currentNM()->mkConst<bool>(false);
+  }
+
+  /**
+   * Destructor
+   * */
+  ~SymmetryDetect() {}
+
+  /** Get the final partition after symmetry detection.
+   *  If a vector in parts contains two variables x and y,
+   *  then assertions and assertions { x -> y, y -> x } are
+   *  equisatisfiable.
+   * */
+  void getPartition(std::vector<std::vector<Node> >& parts, std::vector<Node>& assertions);
```

Comment:
Assertions is not modified, so it should be const here.  I would recommend making this function:

void getPartition(const std::vector<Node>& assertions, std::vector<std::vector<Node> >& parts);

Reply:
What if later we do optimizations on assertions, e.g., rewriting to negation normal form, are we going to modify it or create a new copy of it?

Reply:
Its simpler if we keep this module to *detect* symmetries only.  It will eventually be used by a preprocessing pass that will modify the assertions.